src/Cardano/Wallet/CoinSelection/Fee.hs

```diff
@@ -341,3 +347,96 @@ computeFee (CoinSelection inps outs chgs) =
     -- by definition, impossible; unless we messed up real hard.
     collapse [] _ =
         invariant "outputs are bigger than inputs" (undefined) (const False)
+
+-- | A linear equation on the transaction size. Represents the @\s -> a + b*s@
+-- function where @s@ is the transaction size in bytes, @a@ and @b@ are
+-- constant coefficients.
+data TxSizeLinear = TxSizeLinear Double Double
+    deriving (Eq, Show)
+
+cardanoPolicy :: TxSizeLinear
+cardanoPolicy = TxSizeLinear 155381 43.946
+
+-- | Estimate the fee for a transaction that has @inps@ inputs
+-- and @outps@ outputs as coming from coin selection.
+--
+-- NOTE: The average size of @Attributes AddrAttributes@ and
+--       the transaction attributes @Attributes ()@ are both hard-coded
+estimateCardanoFee
+    :: TxSizeLinear
+    -> CoinSelection
+    -> Fee
+estimateCardanoFee (TxSizeLinear a b) (CoinSelection inps outs _) =
+    Fee $ multioutputCorrection + ceiling (a + b*totalPayload)
+    where
+
+        multioutputCorrection :: Word64
+        multioutputCorrection = case length outs of
+            1 -> 0
+            _ -> 1
+
+        -- The size of TxIn is always the same as it contains the hash and index
+        -- see Cardano.Wallet.Binary (encodeTxIn)
+        sizeOfTxIn :: Int
+        sizeOfTxIn = 42
+
+        sizeOfListLen :: Int
+        sizeOfListLen = 1
+
+        sizeOfListBreak :: Int
+        sizeOfListBreak = 1
+
+        -- The size of [TxIn]
+        sizeOfTxIns :: Int
+        sizeOfTxIns =
+            let n = length inps
+            in sizeOfListLen + n*sizeOfTxIn + (n-1)*sizeOfListBreak + (n-1)*2
+
+        -- The size of TxWitness is always the same as it contains data constructor and hash
+        -- see Cardano.Wallet.Binary (encodeTxWitness)
+        sizeOfTxWitness :: Int
+        sizeOfTxWitness = 139
+
+        -- The size of [TxWitness]
+        sizeOfTxWitnesses :: Int
+        sizeOfTxWitnesses =
+            let n = length inps
+            in sizeOfListLen + n*sizeOfTxWitness + (n-1)*sizeOfListBreak
+
+        -- The size of TxOut depends only on the coin value
+        -- see Cardano.Wallet.Binary (encodeTxOut)
+        sizeOfTxOut :: Word64 -> Int
+        sizeOfTxOut =
+            (+77) . fromIntegral . BL.length . toLazyByteString . encodeWord64
+
+        -- The size of [TxOut]
+        sizeOfTxOuts :: Int
+        sizeOfTxOuts =
+            let n = length outs
+                coins = map (getCoin . coin) outs
+            in sizeOfListLen + sum (map sizeOfTxOut coins) + (n-1)*sizeOfListBreak
+
+
+        -- The size of [[TxIn], [TxWitness]]
+        sizeOfTx :: Int
+        sizeOfTx =
+            sizeOfListLen + sizeOfTxWitnesses + sizeOfTxIns + sizeOfTxOuts + 2*sizeOfListBreak
+
+
+        totalPayload :: Double
+        totalPayload =
+            let n = length outs
+            in fromIntegral $
+               sizeOfListLen + sizeOfListBreak + sizeOfTx + sizeOfTxAttr + n*(sizeOfAddrAttr + sizeOfSignatureWitness) - (n-1)*6
```

Comment:
(n-1)*6 is added artificially here - I cannot explain it at the moment. Just added in order to obtain the same results as with cardano-sl